### PR TITLE
xorg.*: update, add update script

### PR DESCRIPTION
**Pull request by sdht0**
## Description of changes

* Added an update script. Looks for updated versions and commits any changes. Reuses the existing perl script to regenerate nix expressions. Adapted from https://gist.github.com/NickCao/fa592237f768c32ff26688d99b20f81f by @NickCao.
* Updated packages to latest available versions using the script.
* Successfully built all packages. 
* xf86videosavage now builds so removed the broken flag.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by lolbinarycat** [comment]
> Successfully built 4-5 of the packages.

do you mean 4/5? or are you unsure how many packages you successfully built?

**Comment by sdht0** [comment]
> do you mean 4/5? or are you unsure how many packages you successfully built?

Ah sorry for being imprecise. I tried building the updated packages at random, with some failing to build which I then fixed.

**Comment by sdht0** [comment]
@marsam gentle ping, since you merged the earlier xorg PRs.

**Comment by sdht0** [comment]
Updated to add latest releases.

Of note: xorg-server has security updates: https://lists.x.org/archives/xorg-announce/2024-April/003497.html

If needed, xorg-server can be updated in a separate PR directly to master.

cc @vcunat

**Review by MostAwesomeDude — APPROVED** [review]
Obviously up to @vcunat as to merge timing, but this all looks decent. I appreciate the update script!

**Comment by vcunat** [comment]
I'm not aware of any need to control merge timing.

File: pkgs/servers/x11/xorg/update.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+if len(changes) == 0:
+    print("No updates found")
+    exit()
+
+print("Updating tarballs.list...")
+
+with open("./tarballs.list", "w") as f:
+    for l in updated_tarballs:
+        f.write(f'{l}\n')
```
Comment (marsam) [inline pkgs/servers/x11/xorg/update.py:107]: nitpick
```suggestion
    f.writelines(tarball + '\n' for tarball in updated_tarballs)
```

File: pkgs/servers/x11/xorg/update.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env nix-shell
+#!nix-shell --pure --keep NIX_PATH -i python3 -p nix git python3Packages.packaging python3Packages.beautifulsoup4 python3Packages.requests
```
Comment (marsam) [inline pkgs/servers/x11/xorg/update.py:2]: nitpick
```suggestion
#!nix-shell --pure --keep NIX_PATH -i python3 -p nix git "python3.withPackages (ps: [ ps. packaging ps.beautifulsoup4 ps.requests ])"
```

File: pkgs/servers/x11/xorg/update.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+with open("./tarballs.list") as f:
+    lines_tarballs = f.readlines()
+
+updated_tarballs = []
+changes = {}
+changes_text = []
+for line in lines_tarballs:
+    line = line.rstrip("\n")
+    if any(line.startswith(frag) for frag in [mirror, lurl]):
```
Comment (marsam) [inline pkgs/servers/x11/xorg/update.py:79]: I think we can use `continue` here:
```python
    if not line.startswith((mirror, lurl)):
        continue
    pname, rem = line.rsplit("-", 1)
    ...
```

**Review by marsam — COMMENTED** [review]
Looks great! Thank you! 
I left a minor style suggestions. Besides that, LGTM!

File: pkgs/servers/x11/xorg/update.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    url = "https://xorg.freedesktop.org/releases/{}/".format(component)
+    r = requests.get(url)
+    soup = BeautifulSoup(r.text, "html.parser")
+    for a in soup.table.find_all("a"):
+        href = a["href"]
+        if (
+            not href.endswith(".tar.bz2")
+            and not href.endswith(".tar.gz")
+            and not href.endswith(".tar.xz")
+        ):
```
[the remark this answers is not in the document]
Comment (sdht0) [reply pkgs/servers/x11/xorg/update.py]: Ah nice thanks #til

**Comment by sdht0** [comment]
Thanks @marsam. I applied all your suggestions.

File: pkgs/servers/x11/xorg/default.nix [hunk 5]
```diff
@@ -3135,34 +3135,14 @@ self: with self; {
     };
   })) {};
 
-  # THIS IS A GENERATED FILE.  DO NOT EDIT!
-  xf86videonouveau = callPackage ({ stdenv, pkg-config, fetchurl, xorgproto, libdrm, udev, libpciaccess, xorgserver, testers }: stdenv.mkDerivation (finalAttrs: {
```
Comment (marsam) [inline pkgs/servers/x11/xorg/default.nix:3139]: Ofborg is failing to eval because `xf86videonouveau` was removed

Comment (sdht0) [reply pkgs/servers/x11/xorg/default.nix]: Yup that's what I get for not reviewing the diffs after modifying the script.

File: pkgs/servers/x11/xorg/update.py (hunk 3 above)
Comment (sdht0) [reply pkgs/servers/x11/xorg/update.py:79]: This was the bug. Lines should be always added to tarballs file, with or without changes.

**Review by marsam — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**